openssl-dynamic/src/main/c/jnilib.c

```diff
@@ -225,3 +156,275 @@ jint tcn_get_java_env(JNIEnv **env)
+    char* nettyClassName = netty_internal_tcnative_util_prepend(packagePrefix, className);
+    jclass nativeCls = (*env)->FindClass(env, nettyClassName);
+    free(nettyClassName);
+    nettyClassName = NULL;
+    if (nativeCls == NULL) {
+        return JNI_ERR;
+    }
+
+    return (*env)->RegisterNatives(env, nativeCls, methods, numMethods);
+}
+
+static char* netty_internal_tcnative_util_strndup(const char *s, size_t n) {
+// windows does not have strndup
+#ifdef WIN32
+    char* copy = _strdup(s);
+    if (copy != NULL && n < strlen(copy)) {
+        // mark the end
+        copy[n + 1] = '\0';
+    }
+    return copy;
+#else
+    return strndup(s, n);
+#endif
+}
+
+static char* netty_internal_tcnative_util_strstr_last(const char* haystack, const char* needle) {
+    char* prevptr = NULL;
+    char* ptr = (char*) haystack;
+
+    while ((ptr = strstr(ptr, needle)) != NULL) {
+        // Just store the ptr and continue searching.
+        prevptr = ptr;
+        ++ptr;
+    }
+    return prevptr;
+}
+
+/**
+ * The expected format of the library name is "lib<>netty-tcnative" on non windows platforms and "<>netty-tcnative" on windows,
+ *  where the <> portion is what we will return.
+ */
+static char* parsePackagePrefix(const char* libraryPathName, jint* status) {
+    char* packageNameEnd = netty_internal_tcnative_util_strstr_last(libraryPathName, "netty-tcnative");
+    if (packageNameEnd == NULL) {
+        *status = JNI_ERR;
+        return NULL;
+    }
+#ifdef WIN32
+    // on windows we don't have a lib prefix
+    int incr = 0;
+    char* packagePrefix = packageNameEnd;
+#else
+    int incr = 3;
+    char* packagePrefix = netty_internal_tcnative_util_rstrstr(packageNameEnd, libraryPathName, "lib");
+#endif
+
+    if (packagePrefix == NULL) {
+        *status = JNI_ERR;
+        return NULL;
+    }
+    packagePrefix += incr;
+    if (packagePrefix == packageNameEnd) {
+        return NULL;
+    }
+    // packagePrefix length is > 0
+    // Make a copy so we can modify the value without impacting libraryPathName.
+    size_t packagePrefixLen = packageNameEnd - packagePrefix;
+    packagePrefix = netty_internal_tcnative_util_strndup(packagePrefix, packagePrefixLen);
+    // Make sure the packagePrefix is in the correct format for the JNI functions it will be used with.
+    char* temp = packagePrefix;
+    packageNameEnd = packagePrefix + packagePrefixLen;
+    // Package names must be sanitized, in JNI packages names are separated by '/' characters.
+    for (; temp != packageNameEnd; ++temp) {
+        if (*temp == '-') {
+            *temp = '/';
+        }
+    }
+    // Make sure packagePrefix is terminated with the '/' JNI package separator.
+    if(*(--temp) != '/') {
+        temp = packagePrefix;
+        packagePrefix = netty_internal_tcnative_util_prepend(packagePrefix, "/");
+        free(temp);
```

Comment:
this must be wrong, `temp` is modified above (in `--temp`) and then freed. Should be `free(packagePrefix)`, no?